### PR TITLE
Move PSVersionInfo.PSVersionName to first place

### PR DESCRIPTION
**Pull request by iSazonov**
Based on discussion https://github.com/PowerShell/PowerShell/pull/3530#pullrequestreview-32760325

We want `PSVersion` on first place and the rest sorted alphabetically

```powershell
>$PSVersionTable

Name                           Value
----                           -----
PSVersion                      6.0.0-alpha
BuildVersion                   3.0.0.0
CLRVersion
GitCommitId                    v6.0.0-alpha.18-12-gacec58c0491019557864189ebf03324e58874d63-dirty
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSEdition                      Core
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### CODE REVIEW

**Comment by daxian-dbw** [comment]
Quoted from @rkeithhill's [comment](https://github.com/PowerShell/PowerShell/pull/3530#discussion_r111768692)
> Consider putting PSVersion and PSEdition in the first and second positions. After that, I don't care much about the order. I suppose alphabetical after the first two would be fine.

**Comment by iSazonov** [comment]
Ok, tomorrow.

File: src/System.Management.Automation/engine/PSVersionInfo.cs [hunk 1]
```diff
@@ -332,13 +334,22 @@ public override ICollection Keys
...
             {
-                Array arr = new string[base.Keys.Count];
-                base.Keys.CopyTo(arr, 0);
-                Array.Sort(arr, StringComparer.OrdinalIgnoreCase);
-                return arr;
+                ArrayList keyList = new ArrayList(base.Keys);
+                keyList.Sort();
+                MoveItemOnFirstPlace(PSVersionInfo.PSEditionName, keyList);
+                MoveItemOnFirstPlace(PSVersionInfo.PSVersionName, keyList);
+                return keyList.ToArray();
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/PSVersionInfo.cs]: This implementation is a little bit heavy, it involves moving keyList's elements around for 3 times (`sort()` and 2 `MoveItemOnFirstPlace`) + 2 collection allocation (`ArrayList` and `ToArray`).
Maybe the following pseudo code would be better?
```CSharp
Array arr = new string[base.Keys.Count];
int index = 2;
foreach (string key in base.Keys)
{
    switch (key)
       case PSVersionInfo.PSVersionName:
                arr[0] = key;
                break;
       case PSVersionInfo.PSEditionName:
                arr[1] = key;
                break;
       default:
                arr[index++] = key;
                break;
}
Array.Sort(arr, 2, arr.Length - 2; /* the equalityComparer passed in the constructor */);
return arr;
```
It allocate only one collection, and elements in `arr` only get moved around once, within the `Sort` method.

The pseudo code is still not perfect, because we will create an array object every time `Keys` is accessed. If we want to go further to avoid that, we can create our own type that implements `ICollection`and is immutable from outside (no public members to mutate it, just like the type of object that `Hashtable.Keys` returns -- `Hashtable+KeyCollection`). Then we can keep one instance of it with `PSVersionHashTable`, and always reuse and return the same instance from `Keys` property.

File: src/System.Management.Automation/engine/PSVersionInfo.cs [hunk 2]
```diff
@@ -57,7 +59,7 @@ static PSVersionInfo()
             s_psVersionTable = new PSVersionHashTable(StringComparer.OrdinalIgnoreCase);
 
             s_psVersionTable[PSVersionInfo.PSVersionName] = s_psV6Version;
-            s_psVersionTable["PSEdition"] = PSEditionValue;
+            s_psVersionTable[PSVersionInfo.PSEditionName] = PSEditionValue;
             s_psVersionTable["BuildVersion"] = GetBuildVersion();
             s_psVersionTable["GitCommitId"] = GetCommitInfo();
             s_psVersionTable["PSCompatibleVersions"] = new Version[] { s_psV1Version, s_psV2Version, s_psV3Version, s_psV4Version, s_psV5Version, s_psV51Version, s_psV6Version };
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/PSVersionInfo.cs]: Why not have const variables for other keys: "BuildVersion", "GitCommitId" and "PSCompatibleVersions"? 😄

Comment (iSazonov) [inline src/System.Management.Automation/engine/PSVersionInfo.cs]: I did not want to distract the unrelated changes. But I agree to bring it into line if you don't mind.

File: src/System.Management.Automation/engine/PSVersionInfo.cs (hunk 1 above)
Comment (iSazonov) [inline src/System.Management.Automation/engine/PSVersionInfo.cs]: Thanks for help with the code!
I added the optimized sorting and description. 

If we do `Keys` read-only we should do `PSVersionTable` read-only too. And it is a breaking change - currently we can add in `PSVersionTable` own version strings. For example, third-party module can add its strings in `PSVersionTable`. I think it's not worth and  we can live without the `Keys` cache. What are your thoughts?

Comment (daxian-dbw) [inline src/System.Management.Automation/engine/PSVersionInfo.cs]: `PSVersionTable` doesn't need to be read-only for the `Keys` to be read-only. `Hashtable.Keys` returns an object of `System.Collections.Hashtable+KeyCollection`, which expose no public members for you to mutate it.
This would be a pure optimization and is not required for this fix.

**Comment by daxian-dbw** [comment]
@iSazonov, the implementation will break in 2 cases:
1. there are non-string keys -- string array won't work and sort will fail
2. the keys 'PSEdition' or 'PSVersion' get removed from `$psversiontable` -- possible array indexing failure.

I pushed a fix to your branch, and need your help to review the code to see if there are other issues.

**Comment by iSazonov** [comment]
Why we want to allow non-string keys? It seems the keys in `$psversiontable` is NAMEs - other semantics is misguided I believe.

**Comment by iSazonov** [comment]
And maybe move our logic in custom StringComparer?

**Comment by daxian-dbw** [comment]
> Why we want to allow non-string keys? 

We don't want non-string keys, but user can add any key-value pairs to `$psversiontable`. It shouldn't fail when that happens.

**Comment by iSazonov** [comment]
If we are faced with the keys of arbitrary type we would use `LanguagePrimitives.ConvertTo`.

In last commit I also have tried to move sorting logic to custom Comparer. It looks more simple 😄 If you dont agree feel free to revert.

File: src/System.Management.Automation/engine/PSVersionInfo.cs [hunk 3]
```diff
@@ -327,15 +334,47 @@ internal PSVersionHashTable(IEqualityComparer equalityComparer) : base(equalityC
...
+                return keyList;
+            }
+        }
+
+        private class PSVersionTableComparer : Comparer<object>
+        {
+            public override int Compare(object x, object y)
+            {
+                string xString = (string)LanguagePrimitives.ConvertTo(x, typeof(string), null);
+                string yString = (string)LanguagePrimitives.ConvertTo(y, typeof(string), null);
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/PSVersionInfo.cs]: What if `x` or `y` are objects but calling `LanguagePrimitives.ConvertTo` on them returns `PSVersion` or `PSEdition` as the converted string?
I still prefer to do sorting only if keys are all strings. If there is a non-string key, we just give up sorting and return `base.Keys`.
Sorry for being picky with rare cases. Since we are creating a public accessible data structure, we have to consider whatever that might happen.

Comment (iSazonov) [inline src/System.Management.Automation/engine/PSVersionInfo.cs]: This is very useful for me. 😄 The more we did not analyse this deeply before.

All previous options that we had, nothing broke really. The only thing we do here is to change a human output to the screen.
Output (Out-Default) always perform conversion to strings. Even if we have non-string key, it will be converted to a string. And it seems better to even see them sorted.

If a user adds a record with key that is converted to "PSVersion" we see two entries. We could block it but it will be breaking change.
Also we should use `LanguagePrimitives.ConvertTo<string>(Value)` (https://github.com/PowerShell/PowerShell/pull/2678#discussion_r87906295) to avoid "strange" sorting.

```powershell
$a=[pscustomobject]@{v1=123}
$a
$a.ToString() # It is empty!

$PSVersionTable[$a]="asdfgh"
$PSVersionTable

Name                           Value
----                           -----
PSVersion                      6.0.0-alpha
PSEdition                      Core
@{v1=123}                      asdfgh
BuildVersion                   3.0.0.0
CLRVersion
GitCommitId                    v6.0.0-alpha.18-36-gc367cb58d87a84597077fbbe9a5ded5310179878
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0


$b=$a | Add-Member -MemberType ScriptMethod -Name ToString -Value {"PSVersion"} -PassThru -Force
$PSVersionTable[$b]="zxcvbn"
$PSVersionTable

Name                           Value
----                           -----
PSVersion                      zxcvbn
PSVersion                      6.0.0-alpha
PSEdition                      Core
@{v1=123}                      asdfgh
BuildVersion                   3.0.0.0
CLRVersion
GitCommitId                    v6.0.0-alpha.18-36-gc367cb58d87a84597077fbbe9a5ded5310179878
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

Comment (iSazonov) [inline src/System.Management.Automation/engine/PSVersionInfo.cs]: @daxian-dbw Could you please continue?

Comment (daxian-dbw) [inline src/System.Management.Automation/engine/PSVersionInfo.cs]: Sorry @iSazonov, I was busy with some other tasks. I will get back to this one ASAP.

Comment (daxian-dbw) [inline src/System.Management.Automation/engine/PSVersionInfo.cs]: Good argument. I agree to have the `PSVersionTableComparer` for the sorting.

File: src/System.Management.Automation/engine/PSVersionInfo.cs [hunk 4]
```diff
@@ -327,15 +334,47 @@ internal PSVersionHashTable(IEqualityComparer equalityComparer) : base(equalityC
...
-                base.Keys.CopyTo(arr, 0);
-                Array.Sort(arr, StringComparer.OrdinalIgnoreCase);
-                return arr;
+                ArrayList keyList = new ArrayList(base.Keys);
+                keyList.Sort(new PSVersionTableComparer());
+                return keyList;
+            }
+        }
+
+        private class PSVersionTableComparer : Comparer<object>
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/PSVersionInfo.cs]: Maybe just implement the `IComparer` interface?

File: src/System.Management.Automation/engine/PSVersionInfo.cs [hunk 5]
```diff
@@ -327,15 +334,47 @@ internal PSVersionHashTable(IEqualityComparer equalityComparer) : base(equalityC
...
+        {
+            public override int Compare(object x, object y)
+            {
+                string xString = (string)LanguagePrimitives.ConvertTo(x, typeof(string), null);
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/PSVersionInfo.cs]: We should use `CultureInfo.CurrentCulture` for the `formatProvider` parameter.

File: src/System.Management.Automation/engine/PSVersionInfo.cs [hunk 6]
```diff
@@ -327,15 +334,47 @@ internal PSVersionHashTable(IEqualityComparer equalityComparer) : base(equalityC
...
+                string yString = (string)LanguagePrimitives.ConvertTo(y, typeof(string), null);
+                if (PSVersionInfo.PSVersionName == xString)
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/PSVersionInfo.cs]: Instead of `==`, we should use `PSVersionInfo.PSVersionName.Equals(xString, StringComparison.OrdinalIgnoreCase)`

File: src/System.Management.Automation/engine/PSVersionInfo.cs [hunk 7]
```diff
@@ -327,15 +334,47 @@ internal PSVersionHashTable(IEqualityComparer equalityComparer) : base(equalityC
...
+                {
+                    return -1;
+                }
+                else if (PSVersionInfo.PSEditionName == yString)
+                {
+                    return 1;
+                }
+                else
+                {
+                    return xString.CompareTo(yString);
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/PSVersionInfo.cs]: Maybe use `String.Compare(string strA, string strB, System.StringComparison comparisonType)` so that you can specify `StringComparison.OrdinalIgnoreCase`.

**Comment by daxian-dbw** [comment]
#3654 made changes to `PSVersionInfo.cs`, so I'm afraid you need to rebase and resolve the conflicts.

File: src/System.Management.Automation/engine/PSVersionInfo.cs [hunk 8]
```diff
@@ -327,15 +334,47 @@ internal PSVersionHashTable(IEqualityComparer equalityComparer) : base(equalityC
...
         public override ICollection Keys
         {
             get
             {
-                Array arr = new string[base.Keys.Count];
-                base.Keys.CopyTo(arr, 0);
-                Array.Sort(arr, StringComparer.OrdinalIgnoreCase);
-                return arr;
+                ArrayList keyList = new ArrayList(base.Keys);
+                keyList.Sort(new PSVersionTableComparer());
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/PSVersionInfo.cs]: `PSVersionHashTable` can have a static readonly field to hold an instance of `PSVersionTableComparer`, so as to avoid creating a new comparer every time.

File: src/System.Management.Automation/engine/PSVersionInfo.cs (hunk 4 above)
Comment (iSazonov) [inline src/System.Management.Automation/engine/PSVersionInfo.cs]: For PSVersionTable? We only compare keys not PSVersionTables.

File: src/System.Management.Automation/engine/PSVersionInfo.cs (hunk 5 above)
Comment (iSazonov) [inline src/System.Management.Automation/engine/PSVersionInfo.cs]: Fixed.

File: src/System.Management.Automation/engine/PSVersionInfo.cs (hunk 6 above)
Comment (iSazonov) [inline src/System.Management.Automation/engine/PSVersionInfo.cs]: Fixed.

File: src/System.Management.Automation/engine/PSVersionInfo.cs (hunk 7 above)
Comment (iSazonov) [inline src/System.Management.Automation/engine/PSVersionInfo.cs]: Fixed.

File: src/System.Management.Automation/engine/PSVersionInfo.cs (hunk 8 above)
Comment (iSazonov) [inline src/System.Management.Automation/engine/PSVersionInfo.cs]: Fixed.

**Comment by iSazonov** [comment]
@daxian-dbw Conflicts was resolved. Also I moved tests to PSVersionTable.Tests.ps1 file.

File: src/System.Management.Automation/engine/PSVersionInfo.cs (hunk 4 above)
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/PSVersionInfo.cs]: No, for `PSVersionTableComparer`. `ArrayList.Sort(IComparer)` takes an `IComparer` as argument.

File: src/System.Management.Automation/engine/PSVersionInfo.cs [hunk 9]
```diff
@@ -330,15 +338,48 @@ internal PSVersionHashTable(IEqualityComparer equalityComparer) : base(equalityC
...
-                base.Keys.CopyTo(arr, 0);
-                Array.Sort(arr, StringComparer.OrdinalIgnoreCase);
-                return arr;
+                ArrayList keyList = new ArrayList(base.Keys);
+                keyList.Sort(keysComparer);
+                return keyList;
+            }
+        }
+
+        private static PSVersionTableComparer keysComparer = new PSVersionTableComparer();
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/PSVersionInfo.cs]: Please make it `readonly`.
It's recommended to put field declarations before constructors, and for a private static field, it should be named with the prfix `s_` -- `s_keysComparer`. See `s_psVersionTable` in this file for an example.

File: src/System.Management.Automation/engine/PSVersionInfo.cs [hunk 10]
```diff
@@ -330,15 +338,48 @@ internal PSVersionHashTable(IEqualityComparer equalityComparer) : base(equalityC
...
+        private class PSVersionTableComparer : Comparer<object>
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/PSVersionInfo.cs]: Like we discussed, `PSVersionTableComparer` can just implement `IComparer` due to `ArrayList.Sort(IComparer comparer)`.

File: src/System.Management.Automation/engine/PSVersionInfo.cs [hunk 11]
```diff
@@ -5,6 +5,7 @@
 using System.Diagnostics;
 using System.Reflection;
 using System.Collections;
+using System.Collections.Generic;
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/PSVersionInfo.cs]: Once you make `PSVersionTableComparer` implements `IComparer` instead of deriving from `Comparer<object>`, this using directive can be removed.

Comment (iSazonov) [inline src/System.Management.Automation/engine/PSVersionInfo.cs]: Removed.

File: src/System.Management.Automation/engine/PSVersionInfo.cs (hunk 9 above)
Comment (iSazonov) [inline src/System.Management.Automation/engine/PSVersionInfo.cs]: Ready.

File: src/System.Management.Automation/engine/PSVersionInfo.cs (hunk 10 above)
Comment (iSazonov) [inline src/System.Management.Automation/engine/PSVersionInfo.cs]: Fixed.

**Review by daxian-dbw — APPROVED** [review]

**Comment by iSazonov** [comment]
@daxian-dbw Thanks for review and usefull comments!

**Comment by daxian-dbw** [comment]
@iSazonov Thanks for the great fix! 👍
